presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java

```diff
@@ -360,38 +365,85 @@ protected Object visitSearchedCaseExpression(SearchedCaseExpression node, Object
             return result;
         }
 
+        private Object processWithExceptionHandling(Expression expression, Object context)
+        {
+            if (expression == null) {
+                return null;
+            }
+
+            try {
+                return process(expression, context);
+            }
+            catch (RuntimeException e) {
+                // HACK
+                // Certain operations like 0 / 0 or likeExpression may throw exceptions.
+                // Wrap them a FunctionCall that will throw the exception if the expression is actually executed
+                return createFailureFunction(e, type(expression).getTypeSignature().toString());
+            }
+        }
+
+        /**
+         * @return updated WhenClause, the result value if clause is always true, or Optional.empty() if the clause is always false
+         */
+        private Optional<Object> processWhenClause(WhenClause whenClause, Object context, Object nodeOperand, Type operandType)
+        {
+            Object clauseOperand = processWithExceptionHandling(whenClause.getOperand(), context);
+            Object clauseResult = processWithExceptionHandling(whenClause.getResult(), context);
+
+            if (clauseOperand instanceof Expression || nodeOperand instanceof Expression) {
+                return Optional.of(new WhenClause(
+                        toExpression(clauseOperand, type(whenClause.getOperand())),
+                        toExpression(clauseResult, type(whenClause.getResult()))));
+            }
+            if (clauseOperand != null && isEqual(nodeOperand, operandType, clauseOperand, type(whenClause.getOperand()))) {
+                return clauseResult == null ? Optional.of(new NullLiteral()) : Optional.of(clauseResult);
+            }
+            return Optional.empty();
+        }
+
         @Override
         protected Object visitSimpleCaseExpression(SimpleCaseExpression node, Object context)
         {
-            Object operand = process(node.getOperand(), context);
-            if (operand == null) {
-                return node.getDefaultValue().map(defaultValue -> process(defaultValue, context)).orElse(null);
-            }
-            else if (operand instanceof Expression) {
-                return node;
+            Object nodeOperand = processWithExceptionHandling(node.getOperand(), context);
```

Comment:
everywhere else operand is used as the operand of the whenClause, so I renamed this to avoid confusion.